src/lib/utils.ts

```diff
@@ -83,3 +84,24 @@ export const walletCanEditMemo = (wallet: HDWallet): boolean => {
       return true
   }
 }
+
+export type NestedArray = PartialRecord<string, PartialRecord<string, string[]>>
+
+/**
+ * @param data - Partial, two level nested object with array of strings at leaf
+ * @param level1Id - The id into the first level of the nested object
+ * @param level2Id - The id into the second level of the nested object
+ * @param value - The id to union into the deep array
+ *
+ * @returns void - this function mutates data!
+ */
+export const deepUpsertArray = (
+  data: NestedArray,
+  level1Id: string,
+  level2Id: string,
+  value: string,
+): void => {
+  let level1 = data[level1Id]
+  if (!level1) level1 = data[level1Id] = {}
+  level1[level2Id] = union(level1[level2Id] ?? [], [value])
+}
```

Comment:
Might as well make this generic so we can use it on things other than strings:

```suggestion
export type NestedArray<T> = PartialRecord<string, PartialRecord<string, T[]>>

/**
 * @param data - Partial, two level nested object with array of strings at leaf
 * @param level1Id - The id into the first level of the nested object
 * @param level2Id - The id into the second level of the nested object
 * @param value - The id to union into the deep array
 *
 * @returns void - this function mutates data!
 */
export const deepUpsertArray = <T>(
  data: NestedArray<T>,
  level1Id: string,
  level2Id: string,
  value: T,
): void => {
  let level1 = data[level1Id]
  if (!level1) level1 = data[level1Id] = {}
  level1[level2Id] = union(level1[level2Id] ?? [], [value])
}
```

Reply:
done